### PR TITLE
Added support for DictionaryAttributePrefix and ModelExplorer property in liquid

### LINKED ISSUE #2740: Create Liquid Tag for Validations, Notifications
Right now, when we are using Workflows with custom forms made in a Liquid template we can't use the Model Validation and ValidationSummary to display errors. Same probably goes with Notifications which currently are not displaying. 

related to : https://github.com/OrchardCMS/OrchardCore/issues/2241

@sfmskywalker

### LINKED ISSUE #4426: Forms: Add form controls tag helpers
Here I would like to rethink my issue #4250. While creating forms in Orchard Core I encounter some basic problems.  Forms module exposes basic controls as Wigets but flow editor (even improved by recent PR #4106) is not good enought to position all elements. As it was discussed in other places OC out of box should not contain any fancy editor but only some basic building blocks. In my case Forms Widgets are too hight level and not elastic enough. What I came up with was custom controls Tag Helpers that help me to bind Forms Model and also handle validation errors. Tag Helpers can be accessed from Liquid so at the end I have my Liquid page that can be edited in OC Admin menu and where admin user can freely place controls using tag helpers (so model binding and validation will be hadled). This way without fancy forms editor (just simple liquid) I am able to create any form I want. So my final conculsion is proposition to include such Tag Helpers in Forms module.

Example:

```
<label for="Name">Your name</label>
{% helper "form-input", name:'Name', class:'form-control' %}
{% helper "form-validation", for:'Name', class:'form-control' %}
```

### PR DESCRIPTION
**Pull request by ns8482e**
as per comments  on thread
https://github.com/OrchardCMS/OrchardCore/pull/5094#issuecomment-568150500

Now, in liquid, defining following is possible,

```
{% block "form", action:"Edit", route-contentitemid: contentItem.ContentItemId, route-returnUrl:Request.Query["ReturnUrl"],  method:"post", enctype:"multipart/form-data" %}

{%endblock%}
```

And

```
{% form action:"Edit", route-contentitemid: contentItem.ContentItemId, route-returnUrl:Request.Query["ReturnUrl"],  method:"post", enctype:"multipart/form-data" %}

{%endform%}

```
Also Fixes #4426 
Fixes #2740
/cc @jtkech

### CODE REVIEW

**Comment by jtkech** [comment]
Looks good to me but i would need more time, i will review it asap.

There is another PR #4933 i'm working on and where i tried to improve the tag helper setters, [see my comment](https://github.com/OrchardCMS/OrchardCore/pull/4933#issuecomment-565889812). It is not incompatible with what you did here, but maybe better to wait for this PR in case it will be merged soon, then merge dev in your branch, then i will take the time to review this PR with you.

**Comment by ns8482e** [comment]
Added Support for ModelExplorer property, i.e now properties with asp-* having ModelExplorer datatype can be used in liquid. 

`{% helper "input", for: XXX %}`

File: src/OrchardCore/OrchardCore.DisplayManagement.Liquid/TagHelpers/LiquidTagHelperActivator.cs [hunk 1]
```diff
@@ -78,6 +89,24 @@ public LiquidTagHelperActivator(Type type)
                         {
                             value = v.IsNil() ? null : (bool?)Convert.ToBoolean(v.ToStringValue());
                         }
+                        else if(property.PropertyType == typeof(IDictionary<string, string>)) 
+                        {
+                            IDictionary<string, string> dictValue = (IDictionary<string, string>) getter(h); 
+                            if(!string.IsNullOrWhiteSpace(k))
```
Comment (jtkech) [inline src/OrchardCore/OrchardCore.DisplayManagement.Liquid/TagHelpers/LiquidTagHelperActivator.cs]: Add braces

File: src/OrchardCore/OrchardCore.DisplayManagement.Liquid/TagHelpers/LiquidTagHelperActivator.cs [hunk 2]
```diff
@@ -78,6 +89,24 @@ public LiquidTagHelperActivator(Type type)
...
+                                dictValue[k] = v.ToStringValue();
+                            value = dictValue;
+                        }
+                        else if(property.PropertyType == typeof(IDictionary<string, object>)) 
+                        {
+                            IDictionary<string, object> dictValue = (IDictionary<string, object>) getter(h);
+                            if(!string.IsNullOrWhiteSpace(k))
```
Comment (jtkech) [inline src/OrchardCore/OrchardCore.DisplayManagement.Liquid/TagHelpers/LiquidTagHelperActivator.cs]: idem

File: src/OrchardCore/OrchardCore.DisplayManagement.Liquid/TagHelpers/LiquidTagHelperActivator.cs [hunk 3]
```diff
@@ -78,6 +89,24 @@ public LiquidTagHelperActivator(Type type)
...
+                        else if(property.PropertyType == typeof(IDictionary<string, object>)) 
+                        {
+                            IDictionary<string, object> dictValue = (IDictionary<string, object>) getter(h);
+                            if(!string.IsNullOrWhiteSpace(k)) 
+                                dictValue[k] = v.ToObjectValue();
+                            value = dictValue;
+                        }
+                        else if(property.PropertyType == typeof(Microsoft.AspNetCore.Mvc.ViewFeatures.ModelExpression))
+                        {
+                            value = mp.CreateModelExpression<dynamic>(vd,v.ToStringValue() );
```
Comment (jtkech) [inline src/OrchardCore/OrchardCore.DisplayManagement.Liquid/TagHelpers/LiquidTagHelperActivator.cs]: add a space after the comma, remove space before the closing parenthesis

File: src/OrchardCore/OrchardCore.DisplayManagement.Liquid/TagHelpers/LiquidTagHelperActivator.cs [hunk 4]
```diff
@@ -123,18 +152,26 @@ public LiquidTagHelperActivator(Type type)
             {
                 tagHelper = _activatorByFactory(factory, context);
             }
+            
+            var dictKeySeperator =new char[] {'-','_'};
+            
+            var expresionProvider = context.HttpContext.RequestServices.GetService(typeof(Microsoft.AspNetCore.Mvc.ViewFeatures.ModelExpressionProvider)) as
```
Comment (jtkech) [inline src/OrchardCore/OrchardCore.DisplayManagement.Liquid/TagHelpers/LiquidTagHelperActivator.cs]: Add an using Microsoft.AspNetCore.Mvc.ViewFeatures

File: src/OrchardCore/OrchardCore.DisplayManagement.Liquid/TagHelpers/LiquidTagHelperActivator.cs [hunk 5]
```diff
@@ -123,18 +152,26 @@ public LiquidTagHelperActivator(Type type)
...
             foreach (var name in arguments.Names)
             {
                 var propertyName = name.ToPascalCaseUnderscore();
-
+                var dictPropertyName = propertyName.LastIndexOfAny(dictKeySeperator) > -1 ? propertyName.Substring(0,propertyName.LastIndexOfAny(dictKeySeperator)+1) : String.Empty;
+                var dictPropertyKey = propertyName.LastIndexOfAny(dictKeySeperator) > -1 ? propertyName.Substring(propertyName.LastIndexOfAny(dictKeySeperator)+1) : String.Empty;
                 var found = false;
 
-                if (_setters.TryGetValue(propertyName, out var setter))
+                if (_setters.TryGetValue(propertyName, out var setter) || ( !string.IsNullOrWhiteSpace(dictPropertyName) && _setters.TryGetValue(dictPropertyName, out setter) ) )
```
Comment (jtkech) [inline src/OrchardCore/OrchardCore.DisplayManagement.Liquid/TagHelpers/LiquidTagHelperActivator.cs]: remove spaces after / before parenthesis

File: src/OrchardCore/OrchardCore.DisplayManagement.Liquid/TagHelpers/LiquidTagHelperActivator.cs [hunk 6]
```diff
@@ -172,12 +209,28 @@ public static ITagHelper CreateTagHelper(ITagHelperFactory tagHelperFactory, Vie
             var setterDelegate = setterClosedGenericMethod.CreateDelegate(typeof(Action<object, object>), setterAsAction);
 
             return (Action<object, object>)setterDelegate;
-        }
+        }
```
Comment (jtkech) [inline src/OrchardCore/OrchardCore.DisplayManagement.Liquid/TagHelpers/LiquidTagHelperActivator.cs]: remove extra spaces

File: src/OrchardCore/OrchardCore.Mvc.Core/Utilities/StringExtensions.cs [hunk 7]
```diff
@@ -465,6 +466,8 @@ public static string ToPascalCase(this string attribute, char upperAfterDelimite
                 nextIsUpper = false;
             }
 
+            if(endsWithDelimiter)
```
Comment (jtkech) [inline src/OrchardCore/OrchardCore.Mvc.Core/Utilities/StringExtensions.cs]: Add braces to the if block

**Comment by jtkech** [comment]
@ns8482e just did a little review only about formatting ;)

I assume you tried it, does it work as expected?

File: src/OrchardCore/OrchardCore.DisplayManagement.Liquid/TagHelpers/LiquidTagHelperActivator.cs (hunk 1 above)
Comment (ns8482e) [reply src/OrchardCore/OrchardCore.DisplayManagement.Liquid/TagHelpers/LiquidTagHelperActivator.cs]: done

File: src/OrchardCore/OrchardCore.DisplayManagement.Liquid/TagHelpers/LiquidTagHelperActivator.cs (hunk 2 above)
Comment (ns8482e) [reply src/OrchardCore/OrchardCore.DisplayManagement.Liquid/TagHelpers/LiquidTagHelperActivator.cs]: done

**Comment by ns8482e** [comment]
@jtkech yes, It works - However I'll add some sample in demo module and update the documentation too.

**Comment by ns8482e** [comment]
@jtkech Added CRUD sample that uses liquid template in OrchardCore.Demo

**Comment by jtkech** [comment]
@ns8482e okay cool ;)

I will take a look tomorrow

File: src/OrchardCore.Modules/OrchardCore.Demo/Controllers/TodoController.cs [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+
+                    storeModel = model;
+                }
+                else
+                {
+                    storeModel.Text = model.Text;
+                    storeModel.DueDate = model.DueDate;
+                    storeModel.IsCompleted = model.IsCompleted;
+                }
+                _session.Save(storeModel);
```
Comment (Skrypt) [inline src/OrchardCore.Modules/OrchardCore.Demo/Controllers/TodoController.cs]: It should normally use the ContentManager to trigger the proper events.

File: src/OrchardCore.Modules/OrchardCore.Demo/Controllers/TodoController.cs [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+        }
+
+
+        public async Task<IActionResult> Delete(string id)
+        {
+            var list = (await _session.Query<TodoModel>().ListAsync()).ToList();
+            var storeModel = list.Where(c => c.Id == id).FirstOrDefault();
+            if (storeModel != null)
+            {
+                _session.Delete(storeModel);
```
Comment (Skrypt) [inline src/OrchardCore.Modules/OrchardCore.Demo/Controllers/TodoController.cs]: ContentManager

File: src/OrchardCore.Modules/OrchardCore.Demo/Controllers/TodoController.cs [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+                var storeModel = list.Where(c => c.Id == model.Id).FirstOrDefault();
+                if (storeModel == null)
+                {
+                    if (list.Count == 0)
+                    {
+                        model.Id = "1";
+                    }
+                    else
+                    {
+                        model.Id = (list.Select(c => int.Parse(c.Id)).Max() + 1).ToString();
```
Comment (Skrypt) [inline src/OrchardCore.Modules/OrchardCore.Demo/Controllers/TodoController.cs]: This should be handled by the db transaction without requiring to assing an id here.

File: src/OrchardCore.Modules/OrchardCore.Demo/Controllers/TodoController.cs (hunk 8 above)
Comment (ns8482e) [reply src/OrchardCore.Modules/OrchardCore.Demo/Controllers/TodoController.cs]: Goal of this sample is to show how CRUD operation can be done using liquid with Dictonary support and ModelExplorer support in this PR . So didn't use ContentManager of OC

File: src/OrchardCore.Modules/OrchardCore.Demo/Controllers/TodoController.cs (hunk 10 above)
Comment (ns8482e) [reply src/OrchardCore.Modules/OrchardCore.Demo/Controllers/TodoController.cs]: Not using any  content management features

**Comment by ns8482e** [comment]
@jtkech @agriffard Updated document for liquid

**Comment by sebastienros** [comment]
Fixes #4426

**Comment by agriffard** [comment]
@jtkech Can you please review this PR?

File: src/OrchardCore/OrchardCore.DisplayManagement.Liquid/TagHelpers/LiquidTagHelperActivator.cs [hunk 11]
```diff
@@ -58,11 +57,23 @@ public LiquidTagHelperActivator(Type type)
                     {
                         allNames.Add(htmlAttribute.Name.Substring(4).ToPascalCaseDash());
                     }
+                    var dictonaryPrefix = htmlAttribute.DictionaryAttributePrefix;
+                    if (dictonaryPrefix != null)
+                    {
+                        allNames.Add(dictonaryPrefix.ToPascalCaseDash());
+
+                        if (dictonaryPrefix.StartsWith("asp-", StringComparison.Ordinal))
```
Comment (jtkech) [inline src/OrchardCore/OrchardCore.DisplayManagement.Liquid/TagHelpers/LiquidTagHelperActivator.cs]: We could use a `private const string AspPrefix = "asp-";` and then `AspPrefix.Length`.
And use it also above when we do `htmlAttribute.Name.StartsWith("asp-", ...)`

File: src/OrchardCore/OrchardCore.DisplayManagement.Liquid/TagHelpers/LiquidTagHelperActivator.cs [hunk 12]
```diff
@@ -58,11 +57,23 @@ public LiquidTagHelperActivator(Type type)
                     {
                         allNames.Add(htmlAttribute.Name.Substring(4).ToPascalCaseDash());
                     }
+                    var dictonaryPrefix = htmlAttribute.DictionaryAttributePrefix;
```
Comment (jtkech) [inline src/OrchardCore/OrchardCore.DisplayManagement.Liquid/TagHelpers/LiquidTagHelperActivator.cs]: Add a new line before

File: src/OrchardCore/OrchardCore.Mvc.Core/Utilities/StringExtensions.cs [hunk 13]
```diff
@@ -465,6 +466,10 @@ public static string ToPascalCase(this string attribute, char upperAfterDelimite
                 nextIsUpper = false;
             }
 
+            if (endsWithDelimiter)
+            {
+                result.Append(upperAfterDelimiter);
+            }
             return result.ToString();
```
Comment (jtkech) [inline src/OrchardCore/OrchardCore.Mvc.Core/Utilities/StringExtensions.cs]: new line before

File: src/OrchardCore.Modules/OrchardCore.Demo/Controllers/TodoController.cs [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+                    storeModel = model;
+                }
+                else
+                {
+                    storeModel.Text = model.Text;
+                    storeModel.DueDate = model.DueDate;
+                    storeModel.IsCompleted = model.IsCompleted;
+                }
+                _session.Save(storeModel);
+                await _session.CommitAsync();
```
Comment (jtkech) [inline src/OrchardCore.Modules/OrchardCore.Demo/Controllers/TodoController.cs]: You can remove this line and if you can retry it
Normally this is done by the infrastructure

File: src/OrchardCore.Modules/OrchardCore.Demo/Controllers/TodoController.cs [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+
+
+        public async Task<IActionResult> Delete(string id)
+        {
+            var list = (await _session.Query<TodoModel>().ListAsync()).ToList();
+            var storeModel = list.Where(c => c.Id == id).FirstOrDefault();
+            if (storeModel != null)
+            {
+                _session.Delete(storeModel);
+                await _session.CommitAsync();
```
Comment (jtkech) [inline src/OrchardCore.Modules/OrchardCore.Demo/Controllers/TodoController.cs]: Idem, you can remove this line

File: src/OrchardCore.Modules/OrchardCore.Demo/Views/Todo/Edit.liquid [hunk 16]
```diff
@@ -0,0 +1,2 @@
+<h1>Add/Edit Todo (Liquid)</h1>
+{{ Model | shape_render}}
```
Comment (jtkech) [inline src/OrchardCore.Modules/OrchardCore.Demo/Views/Todo/Edit.liquid]: new line at the end of the file

**Comment by jtkech** [comment]
@ns8482e 

I did a review

Could you merge the last dev in your branch and then redo some tests as we have also updated `LiquidTagHelperActivator` through another PR that has been merged.

Then i will approve this PR

File: src/OrchardCore.Modules/OrchardCore.Demo/Controllers/TodoController.cs (hunk 14 above)
Comment (ns8482e) [reply src/OrchardCore.Modules/OrchardCore.Demo/Controllers/TodoController.cs]: Done

File: src/OrchardCore.Modules/OrchardCore.Demo/Views/Todo/Edit.liquid (hunk 16 above)
Comment (ns8482e) [reply src/OrchardCore.Modules/OrchardCore.Demo/Views/Todo/Edit.liquid]: done

File: src/OrchardCore/OrchardCore.Mvc.Core/Utilities/StringExtensions.cs (hunk 13 above)
Comment (ns8482e) [reply src/OrchardCore/OrchardCore.Mvc.Core/Utilities/StringExtensions.cs]: Added new line before

File: src/OrchardCore/OrchardCore.DisplayManagement.Liquid/TagHelpers/LiquidTagHelperActivator.cs (hunk 11 above)
Comment (ns8482e) [reply src/OrchardCore/OrchardCore.DisplayManagement.Liquid/TagHelpers/LiquidTagHelperActivator.cs]: Added private const and replaced at related location

**Comment by ns8482e** [comment]
> @ns8482e
> 
> I did a review
> 
> Could you merge the last dev in your branch and then redo some tests as we have also updated `LiquidTagHelperActivator` through another PR that has been merged.
> 
> Then i will approve this PR

Works - No error

**Comment by jtkech** [comment]
@ns8482e 

I'm working on it and i did some updates that i will commit tomorrow as i have to use the web interface file by file (would be good to add my github user to your repo if not already done).

Here some comments on the main changes

- When storing the `TodoModel : ShapeViewModel` you are saving in the database all the shape properties, so i separated things with a `TodoModel` and a `TodoViewModel `.

- For the dictionary names and keys i don't use anymore the `ToPascal` conversions so i reverted the change in the related extension, and now we only check for a last underscore. So it is now as other liquid attributes that are intended to use underscores e.g `route_returnurl`.

- So now we call `LastIndexOf('_')` once in place of `LastIndexOfAny()` 4 times without having to create a `new char[] { '-', '_' }` (that could have been a static field).

I will commit them tomorrow so that you will be able to see them more in details and do a review.

**Comment by ns8482e** [comment]
@jtkech you removed LastIndexOfAny so I guess route-todoid will not longer work as now it’s only looking for _ 

You may need to update docs for the same

**Comment by jtkech** [comment]
@ns8482e 

Yes for perf reasons and it simplified the code, i already updated the doc accordingly e.g to use `route_returnurl`, and i did again all the tests.

**Comment by jtkech** [comment]
Forgot to say that i changed the `Id` property to a `TodoId` property because `Id` is a special property that is checked on yessql side and that somewhere may be related to the `Id` column.

Also having an identifier relying on the position of the list may create conflicts after removing / adding items. So i used an `IIdgenerator`, the same that we use e.g for `ContentItem`.

File: src/docs/reference/modules/Liquid/README.md [hunk 17]
```diff
@@ -696,30 +696,93 @@ The content of this block can then be reused from the Layout using the `{{ Model
...
+
+In following example, `route_returnUrl` adds `returnUrl` to form action.
+
+```liquid
+{% form action: "Update", method: "post",  route_returnUrl: Request.Query["returnurl"] %}
+    ... ... ...
+{% endform %}
+```
+
+In following example, `route-todoid` adds `Model.TodoId` to hyperlink.
```
Comment (ns8482e) [inline src/docs/reference/modules/Liquid/README.md]: route_todoid

File: src/docs/reference/modules/Liquid/README.md [hunk 18]
```diff
@@ -696,30 +696,93 @@ The content of this block can then be reused from the Layout using the `{{ Model
...
+
+```liquid
+{% a action: "Delete" , controller: "Todo", class: "btn btn-danger", route-todoid: Model.TodoId %}
```
Comment (ns8482e) [inline src/docs/reference/modules/Liquid/README.md]: route_todoid

File: src/docs/reference/modules/Liquid/README.md [hunk 19]
```diff
@@ -774,10 +774,10 @@ In following example, `route_returnUrl` adds `returnUrl` to form action.
 {% endform %}
 ```
 
-In following example, `route-todoid` adds `Model.TodoId` to hyperlink.
+In following example, `route_todoid` adds `Model.TodoId` to hyperlink.
 
 ```liquid
-{% a action: "Delete" , controller: "Todo", class: "btn btn-danger", route-todoid: Model.TodoId %}
+{% a action: "Delete" , controller: "Todo", class: "btn btn-danger", route_todoid: Model.TodoId %}
```
Comment (jtkech) [inline src/docs/reference/modules/Liquid/README.md]: Good catch 👍

**Comment by ns8482e** [comment]
@jtkech Any other suggestions or concerns for this PR ?

**Comment by jtkech** [comment]
@ns8482e 

> Any other suggestions or concerns for this PR ?

LGTM but because `LiquidTagHelperActivator` is a sensitive array about performance, i was waiting for a feedback from @sebastienros. That said, because of [this line of code](https://github.com/OrchardCMS/OrchardCore/blob/94167decff0c7a2f939c1d6eac9f0246292f7634/src/OrchardCore/OrchardCore.DisplayManagement.Liquid/TagHelpers/LiquidTagHelperActivator.cs#L176) we check if we already have a setter before doing extra work, so i think that there is not so much impact on the performance.

**Pull request merged**